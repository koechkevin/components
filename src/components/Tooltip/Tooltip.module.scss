@import "../../styles/global";

.tooltip {
  :global(.ant-tooltip-inner) {
    font-size: 12px;
    background-color: $color-base-dark;
  }
}

.light {
  :global {
    .ant-tooltip-inner {
      font-size: 12px;
      color: $color-base-dark;
      background-color: $color-white;
      box-shadow: 0 2px 4px rgba($color: $color-black, $alpha: 0.16);
    }

    .ant-tooltip-arrow::before {
      background-color: $color-white;
    }
  }
}