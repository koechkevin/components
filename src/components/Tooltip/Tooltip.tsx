import { Tooltip as AntdTooltip } from 'antd';
import { AbstractTooltipProps, RenderFunction } from 'antd/lib/tooltip';
import React, { FC, ReactNode } from 'react';

import styles from './Tooltip.module.scss';

interface Props extends AbstractTooltipProps {
  delay?: number;
  light?: boolean;
  title?: RenderFunction | ReactNode;
}

export const Tooltip: FC<Props> = (props) => {
  const { title, light, children, delay, ...restProps } = props;
  const isTouchDevice = () => 'ontouchstart' in window;
  const overlay = light ? styles.light : styles.tooltip;

  return (
    <AntdTooltip
      mouseEnterDelay={delay}
      mouseLeaveDelay={delay}
      title={!isTouchDevice() && title}
      overlayClassName={overlay}
      {...restProps}
    >
      {children}
    </AntdTooltip>
  );
};

Tooltip.defaultProps = {
  delay: 0.0001,
};

export default Tooltip;
