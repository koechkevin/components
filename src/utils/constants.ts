export enum ProductType {
  Company = 'company-portal',
  Candidate = 'candidate-portal',
}

export enum ChatStatus {
  online = 'online',
  offline = 'offline',
}

export enum FileExtension {
  pdf = 'pdf',
  doc = 'doc',
  docx = 'docx',
  peng = 'png',
  jpg = 'jpg',
  png = 'png',
  tiff = 'tiff',
  xls = 'xls',
  xlsx = 'xlsx',
  page = 'page',
  mp4 = 'mp4',
  mov = 'mov',
  text = 'text',
}

export enum BotAvatarType {
  CalenderYellow = 'CalenderYellow',
  CalenderCyan = 'CalenderCyan',
  CalenderRed = 'CalenderRed',
  CalenderGreen = 'CalenderGreen',
  ClipYellow = 'ClipYellow',
  ClipCyan = 'ClipCyan',
  ClipRed = 'ClipRed',
  ClipGreen = 'ClipGreen',
  PenYellow = 'PenYellow',
  PenCyan = 'PenCyan',
  PenRed = 'PenRed',
  PenGreen = 'PenGreen',
  LockYellow = 'LockYellow',
  LockCyan = 'LockCyan',
  LockRed = 'LockRed',
  LockGreen = 'LockGreen',
}

// API request error code message
export const ErrorCodeMessage = {
  200: 'Request succeeded',
  201: 'Resource created.',
  202: 'A request has entered the background queue (asynchronous task)',
  204: 'No Content - Request succeeded, but no response body',
  400: 'Bad Request - Could not parse request',
  401: 'Unauthorized - No authentication credentials provided or authentication failed',
  403: 'Forbidden - Authenticated user does not have access',
  404: 'Not Found - Resource not found',
  406: 'The format of the request is not available',
  410: 'The requested resource is permanently deleted and will not be retrieved',
  415: 'Unsupported Media Type - POST/PUT/PATCH request occurred without a "application/json" content type',
  422: 'Data validation failed',
  500: 'An internal server error occurred',
  502: 'Gateway error',
  503: 'Service is unavailable, server is temporarily overloaded or maintained',
  504: 'Gateway timeout',
};

export enum RolesConstants {
  AuroraAdmin = 'aurora-admin',
  Candidate = 'candidate',
  CompanyAdmin = 'company-admin',
  CompanyRecruiter = 'company-recruiter',
  CompanyMember = 'company-member',
  CompanyGuest = 'company-guest',
  Chatbot = 'chatbot',
  CompanyOwner = 'company-owner',
  CompanyBilling = 'company-billing',
}

export const months: string[] = [
  'January',
  'February',
  'March',
  'April',
  'May',
  'June',
  'July',
  'August',
  'September',
  'October',
  'November',
  'December',
];

export const daysOfTheWeek: string[] = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
