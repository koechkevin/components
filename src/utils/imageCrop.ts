import ReactCrop from 'react-image-crop';

interface Position {
  x: number;
  y: number;
}

/**
 * Get cropped image url
 * @param {HTMLImageElement} image
 * @param {ReactCrop.Crop} crop
 * @returns {Promise<string>} fileUrl
 */
export function getCroppedImg(image: HTMLImageElement, crop: ReactCrop.Crop): Promise<string> {
  let fileUrl: string;
  const canvas: HTMLCanvasElement = document.createElement('canvas');
  const ctx: CanvasRenderingContext2D = canvas.getContext('2d')!;
  const scaleX = image.naturalWidth / image.width;
  const scaleY = image.naturalHeight / image.height;

  canvas.width = crop.width || 0;
  canvas.height = crop.height || 0;

  ctx.drawImage(
    image,
    (crop.x || 0) * scaleX,
    (crop.y || 0) * scaleY,
    (crop.width || 0) * scaleX,
    (crop.height || 0) * scaleY,
    0,
    0,
    crop.width || 0,
    crop.height || 0,
  );

  return new Promise((resolve, reject) => {
    canvas.toBlob((blob: Blob | null) => {
      if (!blob) {
        reject(new Error('Canvas is empty'));
        return;
      }

      window.URL.revokeObjectURL(fileUrl);
      fileUrl = window.URL.createObjectURL(blob);

      resolve(fileUrl);
    }, 'image/png');
  });
}

/**
 *
 * @param e
 * @returns {Position}
 */
export function getClientPos(e: any): Position {
  let pageX;
  let pageY;

  if (e.touches) {
    [{ pageX, pageY }] = e.touches;
  } else {
    ({ pageX, pageY } = e);
  }

  return {
    x: pageX,
    y: pageY,
  };
}
