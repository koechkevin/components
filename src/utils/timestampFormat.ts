import { differenceInMinutes, endOfDay, format, isSameDay, startOfDay, subDays } from 'date-fns';

const isToday = (t1: Date, t0: Date) => {
  const start = startOfDay(t0);
  const end = endOfDay(t0);

  return isSameDay(t1, start) && isSameDay(t1, end);
};

const isYesterDay = (t1: Date, t0: Date) => {
  const yesterday = subDays(t0, 1);

  const start = startOfDay(yesterday);
  const end = endOfDay(yesterday);

  return isSameDay(t1, start) && isSameDay(t1, end);
};

export const dateFormat = (d1: Date, d0: Date = new Date()) => {
  if (isToday(d1, d0)) {
    return 'Today';
  }

  if (isYesterDay(d1, d0)) {
    return 'Yesterday';
  }

  return format(d1, 'EEEE, d LLL uuuu');
};

export const shortTimeStampFormat = (t: Date = new Date()) => {
  return format(t, 'h:mm a');
};

export const timeStampFormat = (t1: Date, t0: Date = new Date()) => {
  const diffInMinutes = differenceInMinutes(t0, t1);

  if (diffInMinutes < 1) {
    return 'Just now';
  }

  if (diffInMinutes <= 5) {
    return 'A few minutes ago';
  }

  if (diffInMinutes <= 59) {
    return `${diffInMinutes} minutes ago`;
  }

  if (isToday(t1, t0)) {
    return `Today ${format(t1, 'h:mm a')}`;
  }

  if (isYesterDay(t1, t0)) {
    return `Yesterday ${format(t1, 'h:mm a')}`;
  }

  return format(t1, 'd LLL Y, h:mm a');
};

export default timeStampFormat;
