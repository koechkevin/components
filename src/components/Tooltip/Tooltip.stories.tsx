import { number } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import { Button, Col, Row, Typography } from 'antd';
import React from 'react';

import Tooltip from './Tooltip';

const text = <span>Example text</span>;
const buttonWidth = 70;
const { Title } = Typography;

storiesOf('Tooltip', module).add('default', () => {
  const delay = number('Delay', 0.1);

  return (
    <Row gutter={24}>
      <Col lg={12} md={24} style={{ paddingTop: 60, paddingLeft: 32 }}>
        <Title level={3}>Dark Theme</Title>
        <div style={{ marginLeft: buttonWidth, marginTop: 32, whiteSpace: 'nowrap' }}>
          <Tooltip delay={delay} placement="topLeft" title={text}>
            <Button>TL</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="top" title={text}>
            <Button>Top</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="topRight" title={text}>
            <Button>TR</Button>
          </Tooltip>
        </div>
        <div style={{ width: buttonWidth, float: 'left' }}>
          <Tooltip delay={delay} placement="leftTop" title={text}>
            <Button>LT</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="left" title={text}>
            <Button>Left</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="leftBottom" title={text}>
            <Button>LB</Button>
          </Tooltip>
        </div>
        <div style={{ width: buttonWidth, marginLeft: buttonWidth * 4 + 24 }}>
          <Tooltip delay={delay} placement="rightTop" title={text}>
            <Button>RT</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="right" title={text}>
            <Button>Right</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="rightBottom" title={text}>
            <Button>RB</Button>
          </Tooltip>
        </div>
        <div style={{ marginLeft: buttonWidth, clear: 'both', whiteSpace: 'nowrap' }}>
          <Tooltip delay={delay} placement="bottomLeft" title={text}>
            <Button>BL</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="bottom" title={text}>
            <Button>Bottom</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="bottomRight" title={text}>
            <Button>BR</Button>
          </Tooltip>
        </div>
      </Col>
      <Col lg={12} md={24} style={{ paddingTop: 60, paddingLeft: 32 }}>
        <Title level={3}>White Theme</Title>
        <div style={{ marginLeft: buttonWidth, marginTop: 32, whiteSpace: 'nowrap' }}>
          <Tooltip delay={delay} placement="topLeft" title={text} light>
            <Button>TL</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="top" title={text} light>
            <Button>Top</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="topRight" title={text} light>
            <Button>TR</Button>
          </Tooltip>
        </div>
        <div style={{ width: buttonWidth, float: 'left' }}>
          <Tooltip delay={delay} placement="leftTop" title={text} light>
            <Button>LT</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="left" title={text} light>
            <Button>Left</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="leftBottom" title={text} light>
            <Button>LB</Button>
          </Tooltip>
        </div>
        <div style={{ width: buttonWidth, marginLeft: buttonWidth * 4 + 24 }}>
          <Tooltip delay={delay} placement="rightTop" title={text} light>
            <Button>RT</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="right" title={text} light>
            <Button>Right</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="rightBottom" title={text} light>
            <Button>RB</Button>
          </Tooltip>
        </div>
        <div style={{ marginLeft: buttonWidth, clear: 'both', whiteSpace: 'nowrap' }}>
          <Tooltip delay={delay} placement="bottomLeft" title={text} light>
            <Button>BL</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="bottom" title={text} light>
            <Button>Bottom</Button>
          </Tooltip>
          <Tooltip delay={delay} placement="bottomRight" title={text} light>
            <Button>BR</Button>
          </Tooltip>
        </div>
      </Col>
    </Row>
  );
});
