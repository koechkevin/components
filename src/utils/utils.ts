import {
  faFile,
  faFileAlt,
  faFileExcel,
  faFileImage,
  faFilePdf,
  faFileVideo,
  faFileWord,
  IconDefinition,
} from '@fortawesome/pro-regular-svg-icons';
import { pathToRegexp } from 'path-to-regexp';
import urlRegex from 'url-regex';

import { MenuItemProps } from '../components/Menu/interface';
import { FileExtension, RolesConstants } from './constants';

/**
 * Check the array or object if is empty
 * @param {array | object} obj
 */
export function isEmpty(obj: any): boolean {
  return [Object, Array].includes((obj || {}).constructor) && !Object.entries(obj || {}).length;
}

/**
 * Transfer url to url list
 * Example:
 *  '/userinfo/2019/id' => ['/userinfo', '/useinfo/2019, '/userindo/2019/id']
 *
 * @param {string} url
 * @return {string[]}
 */
export function urlToList(url: string): string[] {
  const urlList = url.split('/').filter((i: string) => i);
  return urlList.map((item: string, index) => `/${urlList.slice(0, index + 1).join('/')}`);
}

/**
 * Recursively flatten the data
 * Example:
 *  [{path: string}, {path: string}] => {path1, path2}
 *
 * @param {array} menu
 * @return {string[]}
 */
export function getFlatMenuKeys(menu: MenuItemProps[]): string[] {
  let keys: any[] = [];
  menu.forEach((item: MenuItemProps) => {
    keys.push(item.path);
    if (item.routes) {
      keys = keys.concat(getFlatMenuKeys(item.routes));
    }
  });
  return keys;
}

/**
 * Recursively replace the path
 * Example:
 *  [{path: string}, {path: string}] => {path1, path2}
 *
 * @param {array} menu
 * @return {string[]}
 */
export function replacePath(menu: MenuItemProps[], match: string, value: string): MenuItemProps[] {
  const formated = menu;
  formated.forEach((item: MenuItemProps) => {
    item.path = item.path.replace(match, value);
    if (item.routes) {
      replacePath(item.routes, match, value);
    }
  });
  return formated;
}

/**
 * Get menu matches
 * @param {string[]} flatMenuKeys
 * @return {string[]}
 */
export function getMenuMatches(flatMenuKeys: string[], path: string): string[] {
  const menus = flatMenuKeys.filter((item: string) => {
    if (!item) {
      return [];
    }
    return pathToRegexp(item).test(path);
  });
  return menus;
}

/**
 * Get default collapsed sub menus
 *
 * @param {string} pathname
 * @param {string[]} flatMenuKeys
 * @return {string[]}
 */
export function getDefaultCollapsedSubMenus(pathname: string, flatMenuKeys: string[]): string[] {
  const subMenus = urlToList(pathname)
    .map((item: string) => getMenuMatches(flatMenuKeys, item)[0])
    .filter((item: string) => item)
    .reduce((acc: any, curr: any) => [...acc, curr], ['/']);

  return subMenus;
}

/**
 * Get and concat the first character of user firstName and lastName
 * @param name
 */
export const getNameInitials = (name?: string) => {
  if (name) {
    const arr = name.split(' ');

    if (arr.length >= 2) {
      const str1 = arr[0].length >= 1 ? arr[0][0] : '';
      const str2 = arr[1].length >= 1 ? arr[1][0] : '';
      return `${str1}${str2}`;
    }

    return arr.length >= 1 ? arr[0] : '';
  }

  return '';
};

/**
 * url format
 * @param text
 */
export const urlConverter = (text: string = ''): string => {
  const replaced = text.replace(urlRegex({ strict: false }), (url) => {
    const location = url.indexOf('http') === -1 ? `http://${url}` : url;
    return `<a href="${location}" target="_blank">${url}</a>`;
  });
  return replaced;
};

/**
 * @description Returns the correct icon given an icon type
 * @param {FileExtension} iconType
 * @returns {IconDefinition}
 */
export const selectIcon = (iconType?: FileExtension): IconDefinition => {
  switch (iconType) {
    case FileExtension.pdf:
      return faFilePdf;
    case FileExtension.doc:
    case FileExtension.docx:
      return faFileWord;
    case FileExtension.png:
    case FileExtension.jpg:
    case FileExtension.tiff:
      return faFileImage;
    case FileExtension.xlsx:
    case FileExtension.xls:
      return faFileExcel;
    case FileExtension.mov:
    case FileExtension.mp4:
      return faFileVideo;
    case FileExtension.text:
      return faFileAlt;
    default:
      return faFile;
  }
};

/**
 *
 * @param role
 */
export const resolveRole = (role: RolesConstants | string): string => {
  switch (role) {
    case RolesConstants.AuroraAdmin:
      return 'Admin';
    case RolesConstants.CompanyOwner:
      return 'Owner';
    case RolesConstants.CompanyAdmin:
      return 'Admin';
    case RolesConstants.CompanyRecruiter:
      return 'Recruiter';
    case RolesConstants.CompanyMember:
      return 'Team Member';
    case RolesConstants.CompanyGuest:
      return 'Guest';
    case RolesConstants.CompanyBilling:
      return 'Billing';
    default:
      return role;
  }
};
