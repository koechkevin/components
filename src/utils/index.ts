import { BotAvatarType, ChatStatus, ErrorCodeMessage, FileExtension, ProductType, RolesConstants } from './constants';
import { getClientPos, getCroppedImg } from './imageCrop';
import timestampFormat, { dateFormat, shortTimeStampFormat } from './timestampFormat';
import {
  getDefaultCollapsedSubMenus,
  getFlatMenuKeys,
  getMenuMatches,
  getNameInitials,
  isEmpty,
  replacePath,
  resolveRole,
  selectIcon,
  urlConverter,
  urlToList,
} from './utils';

export {
  isEmpty,
  urlToList,
  dateFormat,
  replacePath,
  ChatStatus,
  BotAvatarType,
  ErrorCodeMessage,
  FileExtension,
  resolveRole,
  ProductType,
  RolesConstants,
  getMenuMatches,
  getFlatMenuKeys,
  getNameInitials,
  timestampFormat,
  shortTimeStampFormat,
  getDefaultCollapsedSubMenus,
  urlConverter,
  selectIcon,
  getCroppedImg,
  getClientPos,
};
