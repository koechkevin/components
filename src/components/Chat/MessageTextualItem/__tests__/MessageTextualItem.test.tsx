import { formatRelative } from 'date-fns';
import { mount } from 'enzyme';
import React from 'react';

import { shortTimeStampFormat } from '../../../../utils';
import { Tooltip } from '../../../Tooltip';
import TextualItem from '../MessageTextualItem';

describe('Tests for MessageItemTextualItem', () => {
  const props = {
    id: 'sampleMessageId',
    createdAt: '2019-10-08 06:22:09',
    updatedAt: '2020-10-08 06:22:09',
    author: {
      id: 'user-fake-id',
      avatarColor: '#4224',
      name: 'Jane Doe',
      productId: 'company-portal',
      chatStatus: 'sample chat status',
    },
    text: 'sample text',
    poppedMenu: <div>Mock Node</div>,
    menuShown: true,
    forwarding: null,
  };

  const wrapper = mount(<TextualItem {...props} />);

  it('should render textual item without crushing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render createdAt timestamp', () => {
    expect(wrapper.contains(shortTimeStampFormat(new Date(props.createdAt)))).toEqual(true);
  });

  it('should render updatedAd timestamp when modified', () => {
    const updatedProps = {
      ...props,
      isModified: true,
    };

    const updatedWrapper = mount(<TextualItem {...updatedProps} />);
    const toolTip = updatedWrapper.find(Tooltip);

    expect(toolTip.props().title).toEqual(formatRelative(new Date(props.updatedAt), new Date()));
    expect(toolTip.find('span').contains('(Edited)')).toEqual(true);
  });
});
