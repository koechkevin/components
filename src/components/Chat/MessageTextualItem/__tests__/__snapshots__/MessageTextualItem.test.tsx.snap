// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for MessageItemTextualItem should render textual item without crushing 1`] = `ReactWrapper {}`;
